### ISSUE TITLE
new ArgParse code does not build with VS2015

### ISSUE THREAD

**Issue by ThiagoIze** [OpenImageIO/oiio#2611]
After commit e81036b6559 (#2531), I can no longer build on Windows using VS2015 because I get build errors of the form:
```
oiio\src\oiiotool\oiiotool.cpp(5244): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(int (__cdecl *)(int,const char **))': cannot convert argument 1 from 'void (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
oiio\src\oiiotool\oiiotool.cpp(5244): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
oiio\src\oiiotool\oiiotool.cpp(5244): note: No constructor could take the source type, or constructor overload resolution was ambiguous
```
 Looking at the code, I can't see how this could be considered ambiguous and this should work, so I suspect a compiler bug. And in fact, if I build with VS2019 I get no error.

I'm hoping very soon we'll be building OIIO with a newer compiler than VS2015, so in my particular case I don't think I need this fixed in OIIO. However, currently the minimum visual studio required, according to the following is 2015:
```
src/include/OpenImageIO/platform.h
# if _MSC_VER < 1900
 # error "This version of OIIO is meant to work only with Visual Studio 2015 or later"
# endif
```
So I think either we'd want to raise the minimum VS (I haven't tested 2017, but I do know 2019 works) or implement a temporary work-around? There's also the possibility I'm messing something up, so of course would be good if someone else could do a sanity check?

**Comment by lgritz** [comment]
Ick, we certainly didn't mean for this little change to have to force a compiler upgrade!

Maybe we can diddle that line into some other form that is equivalent but doesn't trip the error?

**Comment by lgritz** [comment]
I don't see how line oiiotool.cpp:5244 is related to this on either today's master or at commit e81036b.

Can you let me know which commit that error messages is related to, or even better, the line on today's master that trips the error?

Lemme play with some alternatives and see if we can find a workaround.

**Comment by ThiagoIze** [comment]
Sorry about that, I had some debug code. After reverting that hopefully this will line up better.  Just to be sure, the first error I believe corresponds to:

```
    ap.arg("--info")
      .help("Print resolution and basic info on all inputs, detailed metadata if -v is also used (options: format=xml:verbose=1)")
      .action(set_printinfo);
```

here's the full list of errors for e81036b.:
```
oiio\src\oiiotool\oiiotool.cpp(5652): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(int (__cdecl *)(int,const char **))': cannot convert argument 1 from 'void (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
oiio\src\oiiotool\oiiotool.cpp(5652): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
oiio\src\oiiotool\oiiotool.cpp(5652): note: No constructor could take the source type, or constructor overload resolution was ambiguous
oiio\src\oiiotool\oiiotool.cpp(5655): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(int (__cdecl *)(int,const char **))': cannot convert argument 1 from 'void (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
oiio\src\oiiotool\oiiotool.cpp(5655): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
oiio\src\oiiotool\oiiotool.cpp(5655): note: No constructor could take the source type, or constructor overload resolution was ambiguous
oiio\src\oiiotool\oiiotool.cpp(5664): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(int (__cdecl *)(int,const char **))': cannot convert argument 1 from 'void (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
oiio\src\oiiotool\oiiotool.cpp(5664): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
oiio\src\oiiotool\oiiotool.cpp(5664): note: No constructor could take the source type, or constructor overload resolution was ambiguous
oiio\src\oiiotool\oiiotool.cpp(5669): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(int (__cdecl *)(int,const char **))': cannot convert argument 1 from 'void (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
oiio\src\oiiotool\oiiotool.cpp(5669): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
oiio\src\oiiotool\oiiotool.cpp(5669): note: No constructor could take the source type, or constructor overload resolution was ambiguous
oiio\src\oiiotool\oiiotool.cpp(5672): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(int (__cdecl *)(int,const char **))': cannot convert argument 1 from 'void (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
oiio\src\oiiotool\oiiotool.cpp(5672): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
oiio\src\oiiotool\oiiotool.cpp(5672): note: No constructor could take the source type, or constructor overload resolution was ambiguous
oiio\src\oiiotool\oiiotool.cpp(5681): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(int (__cdecl *)(int,const char **))': cannot convert argument 1 from 'void (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
oiio\src\oiiotool\oiiotool.cpp(5681): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
oiio\src\oiiotool\oiiotool.cpp(5681): note: No constructor could take the source type, or constructor overload resolution was ambiguous
oiio\src\oiiotool\oiiotool.cpp(5700): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(int (__cdecl *)(int,const char **))': cannot convert argument 1 from 'void (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
oiio\src\oiiotool\oiiotool.cpp(5700): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
oiio\src\oiiotool\oiiotool.cpp(5700): note: No constructor could take the source type, or constructor overload resolution was ambiguous
oiio\src\oiiotool\oiiotool.cpp(5703): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(int (__cdecl *)(int,const char **))': cannot convert argument 1 from 'void (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
oiio\src\oiiotool\oiiotool.cpp(5703): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
oiio\src\oiiotool\oiiotool.cpp(5703): note: No constructor could take the source type, or constructor overload resolution was ambiguous
oiio\src\oiiotool\oiiotool.cpp(5724): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(int (__cdecl *)(int,const char **))': cannot convert argument 1 from 'void (__cdecl *)(void)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
oiio\src\oiiotool\oiiotool.cpp(5724): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
oiio\src\oiiotool\oiiotool.cpp(5724): note: No constructor could take the source type, or constructor overload resolution was ambiguous
oiio\src\oiiotool\oiiotool.cpp(5725): error C2228: left of '.hidden' must have class/struct/union
```

**Comment by JenusL** [comment]
I'm also stuck at this issue right now with toolset v140
I also have another error in addition to the above ones...
```
oiio\src\include\OpenImageIO/argparse.h(591): error C2075: 'OpenImageIO_v2_2::ArgParse::store_const::<lambda_c281d3b71724d89151a177c99914fd60>::value': array initialization requires a brace-enclosed initializer list
oiio\src\libutil\argparse_test.cpp(181): note: see reference to function template instantiation 'OpenImageIO_v2_2::ArgParse::ArgAction OpenImageIO_v2_2::ArgParse::store_const<char[8]>(const T (&))' being compiled with
          [
              T=char [8]
          ]
```

I also get this warning:
```
oiio\src\libutil\argparse_test.cpp(229): warning C4805: '==': unsafe mix of type 'int' and type 'bool' in operation
```

I found a simple solution to the last error (C2228) that @ThiagoIze reported:
https://github.com/OpenImageIO/oiio/blob/4fc8ce3a29981538fc4fd64b10a8109b2c2cae36/src/oiiotool/oiiotool.cpp#L5309-L5311
The .hidden just needs to be before .action to work, as it's working that way in many other places of the code.
 ```cpp
    ap.arg("--crash")
      .hidden()
      .action(crash_me);
```

The other issues I have no idea how to fix.

**Comment by lgritz** [comment]
@ThiagoIze Can you confirm that changing this one spot to call .hidden() before .action() fixes your error?

@JenusL Can you try changing

      .action(ArgParse::store_const("hey hey"));

to

      .action(ArgParse::store_const(std::string("hey hey")));

Does that help? And if not, how about:

      .action(ArgParse::store_const(OIIO::string_view("hey hey")));

**Comment by ThiagoIze** [comment]
Yes, switching the hidden and action fixes that error.

**Comment by lgritz** [comment]
Do you get a fully clean build now?

I don't understand why that would help -- I looked for maybe a botched declaration on my part that would disallow action() or hidden() from chaining properly, but couldn't find anything wrong. Happy to just switch this order and avoid the compiler bug while we're still supporting that version.

**Comment by ThiagoIze** [comment]
No, it's just that last error (C2228) that this fixes. All the C2664 errors are still there.

**Comment by lgritz** [comment]
Reopening because #2618 doesn't fix all of this.

**Comment by lgritz** [comment]
I started pulling on that ambiguity to try to eliminate it, and ended up with a much more extensive overhaul than I set out to perform. But maybe it helps? @ThiagoIze , can you possibly pull #2619 and see if that compiles cleanly for you?

**Comment by JenusL** [comment]
@lgritz I just tried #2619 and it doesn't work for me.
Pretty much the same errors as before, but in larger quantities:

`oiio\src\igrep\igrep.cpp(150): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(std::function<void (void)> &&)': cannot convert argument 1 from 'int (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'`
```
oiio\src\oiiotool\oiiotool.cpp(5198): error C2664: 'OpenImageIO_v2_2::ArgParse::Arg &OpenImageIO_v2_2::ArgParse::Arg::action(std::function<void (void)> &&)': cannot convert argument 1 from 'int (__cdecl *)(OpenImageIO_v2_2::span<const char *const ,-1>)' to 'OpenImageIO_v2_2::ArgParse::ArgAction &&'
  oiio\src\oiiotool\oiiotool.cpp(5198): note: Reason: cannot convert from 'overloaded-function' to 'OpenImageIO_v2_2::ArgParse::ArgAction'
  oiio\src\oiiotool\oiiotool.cpp(5198): note: No constructor could take the source type, or constructor overload resolution was ambiguous
oiio\src\oiiotool\oiiotool.cpp(5595): fatal error C1003: error count exceeds 100; stopping compilation
```
```
oiio\src\include\OpenImageIO/argparse.h(583): error C2075: 'OpenImageIO_v2_2::ArgParse::store_const::<lambda_c281d3b71724d89151a177c99914fd60>::value': array initialization requires a brace-enclosed initializer list
  oiio\src\libutil\argparse_test.cpp(181): note: see reference to function template instantiation 'OpenImageIO_v2_2::ArgParse::ArgAction OpenImageIO_v2_2::ArgParse::store_const<char[8]>
  (const T (&))' being compiled
          with
          [
              T=char [8]
          ]

```

**Comment by lgritz** [comment]
Oh dear, maybe I've moved in inexactly the wrong direction. :-(

**Comment by lgritz** [comment]
Thanks for trying it, let me tinker a bit more...

**Comment by lgritz** [comment]
Aha, I might have an idea!

**Comment by lgritz** [comment]
Yet another stab at it:  #2620 

Sorry to impose, can you try? I *think* maybe I understand the problem now, might be not correctly figuring out how to turn something into an rvalue.

**Comment by JenusL** [comment]
#2620 didn't change any of the errors compared to master. Still cannot convert void to ArgAction.

**Comment by ThiagoIze** [comment]
Maybe this will help https://gcc.godbolt.org/z/jMpVPQ since you'll be able to test out your ideas directly.

**Comment by ThiagoIze** [comment]
Also, using that URL you can also try VS2017 and see that it fails there as well, while VS2019 passes.

**Comment by JenusL** [comment]
Yeah it's overloading of void that doesn't work...
@lgritz In #2620 you only wrap the overloading of int. Do you have to do the same for void?

**Comment by lgritz** [comment]
I think the problem is that you can't overload on return type, so I have to convert them all to int or all to void.

So, I can build on VS2019 on GitHub CI... I think that 2019 is able to build in 2017 or 2015 mode? If I can just figure out the command line args or cmake flags to make it do that, then I can test this all on my own and not bother you guys again until I've gotten it figured out for real.

**Comment by lgritz** [comment]
Good news -- I was able to coax Appveyor into building with a version of MSVS that exhibits the problem. Stay tuned, now I have a way to test fixes on my own.

**Comment by lgritz** [comment]
Please try #2622. I think I have it now.

**Comment by JenusL** [comment]
YES! #2622 works without any issues. Thank you!

**Comment by ThiagoIze** [comment]
I can also confirm this built for me. Thanks Larry!

**Comment by lgritz** [comment]
Merged #2622
